File: ios/OpenBot/OpenBot/tflite/DetectorDefault.swift
//
// Created by Sparsh Jain on 01/11/22.
//

import Foundation
import TensorFlowLite

/// This TensorFlow Lite detector works with the quantized MobileNet and EfficientDet model.
class DetectorDefault: Detector {

    // outputLocations: array of shape [Batchsize, NUM_DETECTIONS,4]
    // contains the location of detected boxes
    private var outputLocations: UnsafeMutableBufferPointer<Float32>?;
    // outputClasses: array of shape [Batchsize, NUM_DETECTIONS]
    // contains the classes of detected boxes
    private var outputClasses: UnsafeMutableBufferPointer<Float32>?;
    // outputScores: array of shape [Batchsize, NUM_DETECTIONS]
    // contains the scores of detected boxes
    private var outputScores: UnsafeMutableBufferPointer<Float32>?;
    // outputScores: array of shape [Batchsize, NUM_DETECTIONS]
    // contains the scores of detected boxes
    private var numDetections: UnsafeMutableBufferPointer<Float32>?;

    // indices in tflite model
    private var outputLocationsIdx: Int = -1;
    private var outputClassesIdx: Int = -1;
    private var outputScoresIdx: Int = -1;
    private var numDetectionsIdx: Int = -1;

    /// Initializes a DetectorDefault.
    ///
    /// - Parameters:
    ///     - model: the model considered in the inference process
    ///     - device: CPU, GPU or XNNPACK (neural engine)
    ///     - numThreads: number of threads used tin the inference process
    override init(model: Model, device: RuntimeDevice, numThreads: Int) throws {
        try super.init(model: model, device: device, numThreads: numThreads)
    }

    /// Get boolean that determines if aspect ratio should be preserved when rescaling.
    ///
    /// - Returns: true if aspect ratio should be preserved when rescaling.
    override func getMaintainAspect() -> Bool {
        false;
    }

    /// Getter function
    ///
    /// - Returns: path of the file containing the diferent labels
    override func getLabelPath() -> String {
        "labelmap.txt";
    }

    /// Get the number of bytes that is used to store a single color channel value.
    ///
    /// - Returns: The number of bytes used to store a single color channel value.
    override func getNumBytesPerChannel() -> Int {
        1; // the quantized model only uses a single byte
    }

    /// Getter function
    ///
    /// - Returns: number of detections of a given class
    override func getNumDetections() -> Int {
        NUM_DETECTIONS;
    }

    override func parseTFlite() {
        let index0 = try! tflite?.output(at: 0);
        let index1 = try! tflite?.output(at: 1);
        let index2 = try! tflite?.output(at: 2);
        let index3 = try! tflite?.output(at: 3);

        assignIndex(tensor: index0, index: 0);
        assignIndex(tensor: index1, index: 1);
        assignIndex(tensor: index2, index: 2);
        assignIndex(tensor: index3, index: 3);
        NUM_DETECTIONS = try! tflite?.output(at: outputLocationsIdx).shape.dimensions[1] ?? 0;
    }

    /// Index  allocation to browse the tflite object
    ///
    /// - Parameters:
    ///     - tensor: a tensor structure from a tflite object
    ///     - index: the index to be assigned to the tensor
    func assignIndex(tensor: Tensor?, index: Int) {
        switch (tensor?.name) {
        case "TFLite_Detection_PostProcess":
            outputLocationsIdx = index;
            break
        case "TFLite_Detection_PostProcess:1":
            outputClassesIdx = index;
            break
        case "TFLite_Detection_PostProcess:2":
            outputScoresIdx = index;
            break
        case "TFLite_Detection_PostProcess:3":
            numDetectionsIdx = index;
            break
        case "StatefulPartitionedCall:3":
            outputLocationsIdx = index;
            break
        case "StatefulPartitionedCall:2":
            outputClassesIdx = index;
            break
        case "StatefulPartitionedCall:1":
            outputScoresIdx = index;
            break
        case "StatefulPartitionedCall:0":
            numDetectionsIdx = index;
            break
        default:
            break;
        }
    }

    /// Copy data to the input of the neural network
    override func feedData() throws {
        do {

            let outputLocationsTensor = try tflite?.output(at: outputLocationsIdx);
            let outputClassesTensor = try tflite?.output(at: outputClassesIdx);
            let outputScoresTensor = try tflite?.output(at: outputScoresIdx);
            let numDetectionsTensor = try tflite?.output(at: numDetectionsIdx);

            let outputSize = outputLocationsTensor?.shape.dimensions.reduce(1, { x, y in x * y }) ?? 0
            outputLocations = UnsafeMutableBufferPointer<Float32>.allocate(capacity: outputSize)
            _ = outputLocationsTensor?.data.copyBytes(to: outputLocations!);

            let outputClassesTensorSize = outputClassesTensor?.shape.dimensions.reduce(1, { x, y in x * y }) ?? 0
            outputClasses = UnsafeMutableBufferPointer<Float32>.allocate(capacity: outputClassesTensorSize)
            _ = outputClassesTensor?.data.copyBytes(to: outputClasses!);

            let outputScoresTensorSize = outputScoresTensor?.shape.dimensions.reduce(1, { x, y in x * y }) ?? 0
            outputScores = UnsafeMutableBufferPointer<Float32>.allocate(capacity: outputScoresTensorSize)
            _ = outputScoresTensor?.data.copyBytes(to: outputScores!);

            let numDetectionsTensorSize = numDetectionsTensor?.shape.dimensions.reduce(1, { x, y in x * y }) ?? 0
            numDetections = UnsafeMutableBufferPointer<Float32>.allocate(capacity: numDetectionsTensorSize)
            _ = numDetectionsTensor?.data.copyBytes(to: numDetections!);

        } catch {
            print("error:\(error)")
        }
    }

    /// Show the best detections, after scaling them back to the input size.
    ///
    /// - Parameters:
    ///     - className:
    ///     - width:
    ///     - height:
    /// - Returns: an array of "Recognition" objects containing
    override func getRecognitions(className: String, width: Int, height: Int) -> [Recognition] {
        var recognitions: [Recognition] = [];
        for i in 0..<getNumDetections() {
            let xPos = CGFloat(outputLocations![(4 * i) + 1]) * CGFloat(getImageSizeX());
            let yPos = CGFloat(outputLocations![(4 * i)]) * CGFloat(getImageSizeY());
            let w = CGFloat(outputLocations![(4 * i) + 3]) * CGFloat(getImageSizeX()) - xPos;
            let h = CGFloat(outputLocations![(4 * i) + 2]) * CGFloat(getImageSizeY()) - yPos;
            let scaleX = CGFloat(width) / CGFloat(getImageSizeX())
            let scaleY = CGFloat(height) / CGFloat(getImageSizeY())
            //let scale = min(scaleX, scaleY)
            let dx = (CGFloat(width) - scaleX * CGFloat(getImageSizeX())) / 2
            let dy = (CGFloat(height) - scaleY * CGFloat(getImageSizeY())) / 2
            let transform = CGAffineTransform.identity.translatedBy(x: dx, y: dy).scaledBy(x: scaleX, y: scaleY)
            let detection = CGRect(x: xPos, y: yPos, width: w, height: h).applying(transform);
            // SSD Mobilenet V1 Model assumes class 0 is background class
            // in label file and class labels start from 1 to number_of_classes+1,
            // while outputClasses correspond to class index from 0 to number_of_classes
            let classId: Int = Int(outputClasses![i]);
            let labelId: Int = classId + 1;
            if (className == labels[labelId]) {
                recognitions.append(Recognition(id: String(i), title: labels[labelId], confidence: outputScores![i], location: detection, classId: classId));
            }

        }
        // Execute non-maximum suppression
        return nms(recognitions: recognitions);
    }


    override func getMultipleRecognitions(classA: String, classB: String, width: Int, height: Int) -> [Recognition] {
        var recognitions: [Recognition] = [];
        for i in 0..<getNumDetections() {
            let xPos = CGFloat(outputLocations![(4 * i) + 1]) * CGFloat(getImageSizeX());
            let yPos = CGFloat(outputLocations![(4 * i)]) * CGFloat(getImageSizeY());
            let w = CGFloat(outputLocations![(4 * i) + 3]) * CGFloat(getImageSizeX()) - xPos;
            let h = CGFloat(outputLocations![(4 * i) + 2]) * CGFloat(getImageSizeY()) - yPos;
            let scaleX = CGFloat(width) / CGFloat(getImageSizeX())
            let scaleY = CGFloat(height) / CGFloat(getImageSizeY())
            //let scale = min(scaleX, scaleY)
            let dx = (CGFloat(width) - scaleX * CGFloat(getImageSizeX())) / 2
            let dy = (CGFloat(height) - scaleY * CGFloat(getImageSizeY())) / 2
            let transform = CGAffineTransform.identity.translatedBy(x: dx, y: dy).scaledBy(x: scaleX, y: scaleY)
            let detection = CGRect(x: xPos, y: yPos, width: w, height: h).applying(transform);
            // SSD Mobilenet V1 Model assumes class 0 is background class
            // in label file and class labels start from 1 to number_of_classes+1,
            // while outputClasses correspond to class index from 0 to number_of_classes
            let classId: Int = Int(outputClasses![i]);
            let labelId: Int = classId + 1;
            if (classA == labels[labelId] || classB == labels[labelId]) {
                recognitions.append(Recognition(id: String(i), title: labels[labelId], confidence: outputScores![i], location: detection, classId: classId));
            }
        }
        // Execute non-maximum suppression
        return nms(recognitions: recognitions);
    }

    override func getAllRecognitions(width: Int, height: Int) -> [Recognition] {
        var recognitions: [Recognition] = [];
        for i in 0..<getNumDetections() {
            let xPos = CGFloat(outputLocations![(4 * i) + 1]) * CGFloat(getImageSizeX());
            let yPos = CGFloat(outputLocations![(4 * i)]) * CGFloat(getImageSizeY());
            let w = CGFloat(outputLocations![(4 * i) + 3]) * CGFloat(getImageSizeX()) - xPos;
            let h = CGFloat(outputLocations![(4 * i) + 2]) * CGFloat(getImageSizeY()) - yPos;
            let scaleX = CGFloat(width) / CGFloat(getImageSizeX())
            let scaleY = CGFloat(height) / CGFloat(getImageSizeY())
            //let scale = min(scaleX, scaleY)
            let dx = (CGFloat(width) - scaleX * CGFloat(getImageSizeX())) / 2
            let dy = (CGFloat(height) - scaleY * CGFloat(getImageSizeY())) / 2
            let transform = CGAffineTransform.identity.translatedBy(x: dx, y: dy).scaledBy(x: scaleX, y: scaleY)
            let detection = CGRect(x: xPos, y: yPos, width: w, height: h).applying(transform);
            // SSD Mobilenet V1 Model assumes class 0 is background class
            // in label file and class labels start from 1 to number_of_classes+1,
            // while outputClasses correspond to class index from 0 to number_of_classes
            let classId: Int = Int(outputClasses![i]);
            let labelId: Int = classId + 1;
            recognitions.append(Recognition(id: String(i), title: labels[labelId], confidence: outputScores![i], location: detection, classId: classId));
        }
        // Execute non-maximum suppression
        return nms(recognitions: recognitions);
    }
}

File: android/robot/src/main/java/org/openbot/tflite/DetectorDefault.java
/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

// Modified by Matthias Mueller - Intel Intelligent Systems Lab - 2020

package org.openbot.tflite;

import android.app.Activity;
import android.graphics.RectF;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/** This TensorFlow Lite detector works with the quantized MobileNet and EfficientDet model. */
public class DetectorDefault extends Detector {

  // outputLocations: array of shape [Batchsize, NUM_DETECTIONS,4]
  // contains the location of detected boxes
  private float[][][] outputLocations;
  // outputClasses: array of shape [Batchsize, NUM_DETECTIONS]
  // contains the classes of detected boxes
  private float[][] outputClasses;
  // outputScores: array of shape [Batchsize, NUM_DETECTIONS]
  // contains the scores of detected boxes
  private float[][] outputScores;
  // numDetections: array of shape [Batchsize]
  // contains the number of detected boxes
  private float[] numDetections;

  // indices in tflite model
  private int outputLocationsIdx;
  private int outputClassesIdx;
  private int outputScoresIdx;
  private int numDetectionsIdx;

  /**
   * Initializes a {@code DetectorDefault}.
   *
   * @param activity
   */
  public DetectorDefault(Activity activity, Model model, Device device, int numThreads)
      throws IOException {
    super(activity, model, device, numThreads);
  }

  @Override
  public boolean getMaintainAspect() {
    return false;
  }

  @Override
  public RectF getCropRect() {
    return new RectF(0.0f, 0.0f, 0.0f, 0.0f);
  }

  @Override
  protected String getLabelPath() {
    return "networks/labelmap.txt";
  }

  @Override
  protected int getNumBytesPerChannel() {
    // the quantized model uses a single byte only
    return 1;
  }

  @Override
  protected final int getNumDetections() {
    return NUM_DETECTIONS;
  }

  @Override
  protected final void parseTflite() {
    try {
      outputLocationsIdx = tflite.getOutputIndex("TFLite_Detection_PostProcess");
      outputClassesIdx = tflite.getOutputIndex("TFLite_Detection_PostProcess:1");
      outputScoresIdx = tflite.getOutputIndex("TFLite_Detection_PostProcess:2");
      numDetectionsIdx = tflite.getOutputIndex("TFLite_Detection_PostProcess:3");
      NUM_DETECTIONS = tflite.getOutputTensor(outputLocationsIdx).shape()[1];
    } catch (IllegalArgumentException e) {
      outputLocationsIdx = tflite.getOutputIndex("StatefulPartitionedCall:3");
      outputClassesIdx = tflite.getOutputIndex("StatefulPartitionedCall:2");
      outputScoresIdx = tflite.getOutputIndex("StatefulPartitionedCall:1");
      numDetectionsIdx = tflite.getOutputIndex("StatefulPartitionedCall:0");
      NUM_DETECTIONS = tflite.getOutputTensor(outputLocationsIdx).shape()[1];
    }
  }

  @Override
  protected void addPixelValue(int pixelValue) {
    imgData.put((byte) ((pixelValue >> 16) & 0xFF));
    imgData.put((byte) ((pixelValue >> 8) & 0xFF));
    imgData.put((byte) (pixelValue & 0xFF));
  }

  @Override
  protected void runInference() {
    Object[] inputArray = {imgData};
    try {
      tflite.runForMultipleInputsOutputs(inputArray, outputMap);
    }
    catch(Exception e){
      System.out.println("Error Occurred: " + e);
    }
  }

  @Override
  protected void feedData() {
    outputLocations = new float[1][getNumDetections()][4];
    outputClasses = new float[1][getNumDetections()];
    outputScores = new float[1][getNumDetections()];
    numDetections = new float[1];

    outputMap.put(outputLocationsIdx, outputLocations);
    outputMap.put(outputClassesIdx, outputClasses);
    outputMap.put(outputScoresIdx, outputScores);
    outputMap.put(numDetectionsIdx, numDetections);
  }

  @Override
  protected List<Recognition> getRecognitions(String className) {
    // Show the best detections.
    // after scaling them back to the input size.
    final ArrayList<Recognition> recognitions = new ArrayList<>(getNumDetections());
    for (int i = 0; i < getNumDetections(); ++i) {
      final RectF detection =
          new RectF(
              outputLocations[0][i][1] * getImageSizeY(),
              outputLocations[0][i][0] * getImageSizeX(),
              outputLocations[0][i][3] * getImageSizeY(),
              outputLocations[0][i][2] * getImageSizeX());
      // SSD Mobilenet V1 Model assumes class 0 is background class
      // in label file and class labels start from 1 to number_of_classes+1,
      // while outputClasses correspond to class index from 0 to number_of_classes
      int classId = (int) outputClasses[0][i];
      int labelId = classId + 1;
      if (labels.get(labelId).contentEquals(className)) {
        recognitions.add(
            new Recognition("" + i, labels.get(labelId), outputScores[0][i], detection, classId));
      }
    }
    return nms(recognitions);
  }

  /**
   * Retrieves both recognized objects from the model's output and organizes them into ArrayLists
   * based on class names.
   *
   * @return An ArrayList of ArrayLists containing Recognition results for both classes.
   */
  @Override
  protected ArrayList<ArrayList<Recognition>> getMultipleRecognitions(String classNameFirst, String classNameSecond) {
    final ArrayList<ArrayList<Recognition>> multipleRecognitions = new ArrayList<>(getNumDetections());
    multipleRecognitions.add(new ArrayList<>());
    multipleRecognitions.add(new ArrayList<>());
    for (int i = 0; i < getNumDetections(); ++i) {
      final RectF detection =
              new RectF(
                      outputLocations[0][i][1] * getImageSizeY(),
                      outputLocations[0][i][0] * getImageSizeX(),
                      outputLocations[0][i][3] * getImageSizeY(),
                      outputLocations[0][i][2] * getImageSizeX());
      // SSD Mobilenet V1 Model assumes class 0 is background class
      // in label file and class labels start from 1 to number_of_classes+1,
      // while outputClasses correspond to class index from 0 to number_of_classes
      int classId = (int) outputClasses[0][i];
      int labelId = classId + 1;
      if (multipleRecognitions.get(0).isEmpty() && labels.get(labelId).contentEquals(classNameFirst)) {
        multipleRecognitions.get(0).add(
                new Recognition("" + i, labels.get(labelId), outputScores[0][i], detection, classId));
      }
      if (multipleRecognitions.get(1).isEmpty() && labels.get(labelId).contentEquals(classNameSecond)) {
        multipleRecognitions.get(1).add(
                new Recognition("" + i, labels.get(labelId), outputScores[0][i], detection, classId));
      }
    }
    return multipleNMS(multipleRecognitions);
  }

  @Override
  protected ArrayList<Recognition> getAllRecognition() {
    final ArrayList<Recognition> allRecognitions = new ArrayList<>(getNumDetections());
    for (int i = 0; i < getNumDetections(); ++i) {
      final RectF detection =
              new RectF(
                      outputLocations[0][i][1] * getImageSizeY(),
                      outputLocations[0][i][0] * getImageSizeX(),
                      outputLocations[0][i][3] * getImageSizeY(),
                      outputLocations[0][i][2] * getImageSizeX());
      // SSD Mobilenet V1 Model assumes class 0 is background class
      // in label file and class labels start from 1 to number_of_classes+1,
      // while outputClasses correspond to class index from 0 to number_of_classes
      int classId = (int) outputClasses[0][i];
      int labelId = classId + 1;
      allRecognitions.add(
                new Recognition("" + i, labels.get(labelId), outputScores[0][i], detection, classId));
    }
    return nms(allRecognitions);
  }

}
